### PR TITLE
When deploying Next.js utilize esbuild via NPX

### LINKED ISSUE #5268: firebase deploy --only hosting fails with webframeworks experiment enabled
### Environment info

**firebase-tools:**
11.16.1

**Platform:**
macOS arm

### Steps to reproduce

- create Next.js app with `yarn create next-app --typescript`
- initialise firebase hosting
- run `firebase deploy --only hosting`

### Expected behavior

deploy succeeded

### Actual behavior

Deploy fails

<details><summary>firebase-debug.log</summary>

```
[debug] [2022-11-25T15:10:57.874Z] Error: The package "esbuild-darwin-64" could not be found, and is needed by esbuild.

If you are installing esbuild with npm, make sure that you don't specify the
"--no-optional" flag. The "optionalDependencies" package.json feature is used
by esbuild to install the correct binary executable for your current platform.
    at generateBinPath (/Users/lesnitsky/.cache/firebase/tools/lib/node_modules/firebase-tools/node_modules/esbuild/lib/main.js:1786:15)
    at esbuildCommandAndArgs (/Users/lesnitsky/.cache/firebase/tools/lib/node_modules/firebase-tools/node_modules/esbuild/lib/main.js:1850:33)
    at ensureServiceIsRunning (/Users/lesnitsky/.cache/firebase/tools/lib/node_modules/firebase-tools/node_modules/esbuild/lib/main.js:2015:25)
    at Object.build (/Users/lesnitsky/.cache/firebase/tools/lib/node_modules/firebase-tools/node_modules/esbuild/lib/main.js:1906:26)
    at ɵcodegenFunctionsDirectory (/Users/lesnitsky/.cache/firebase/tools/lib/node_modules/firebase-tools/lib/frameworks/next/index.js:190:23)
    at async prepareFrameworks (/Users/lesnitsky/.cache/firebase/tools/lib/node_modules/firebase-tools/lib/frameworks/index.js:302:84)
    at async deploy (/Users/lesnitsky/.cache/firebase/tools/lib/node_modules/firebase-tools/lib/deploy/index.js:53:13)
[error] 
[error] Error: An unexpected error has occurred.
```

</details>

### PR DESCRIPTION
**Pull request by jamesdaniels**
Switch `firebase deploy` Next.js to shell out to `npx esbuild` rather than having esbuild as a (pinned) dependency in firebase-tools. Fixes #5268

### CODE REVIEW

File: src/frameworks/next/index.ts [hunk 1]
```diff
@@ -337,32 +336,28 @@ export async function ɵcodegenPublicDirectory(sourceDir: string, destDir: strin
   }
 }
 
+function extractDeps(ls: any): string[] {
```
Comment (jamesdaniels) [inline src/frameworks/next/index.ts:339]: type

File: src/frameworks/next/index.ts [hunk 2]
```diff
@@ -337,32 +336,28 @@ export async function ɵcodegenPublicDirectory(sourceDir: string, destDir: strin
...
-      absWorkingDir: sourceDir,
-      entryPoints: ["next.config.js"],
-      outfile: join(destDir, "next.config.js"),
-      target: `node${NODE_VERSION}`,
-      platform: "node",
-    });
+    const deps = extractDeps(
+      JSON.parse(execSync(`npm ls --omit=dev --all --json`, { cwd: sourceDir }).toString())
+    );
+    const externals = deps.map((it) => `--external:${it}`).join(" ");
```
Comment (jamesdaniels) [inline src/frameworks/next/index.ts:356]: drop itermediate

File: src/frameworks/next/index.ts [hunk 3]
```diff
@@ -337,32 +336,39 @@ export async function ɵcodegenPublicDirectory(sourceDir: string, destDir: strin
...
 }
 
+interface Dependency {
+  dependencies: {
+    [key: string]: Dependency;
+  };
+}
+
+function extractDeps(ls: Dependency | null): string[] {
+  return Object.keys(ls?.dependencies || {}).reduce(
```
Comment (jamesdaniels) [inline src/frameworks/next/index.ts:346]: can drop ? and ! with an early return

File: src/frameworks/next/index.ts [hunk 4]
```diff
@@ -344,25 +344,23 @@ export async function ɵcodegenFunctionsDirectory(sourceDir: string, destDir: st
...
+      .map((it) => `--external:${it}`)
+      .concat(
+        "--bundle",
+        "--platform=node",
+        `--target=node${NODE_VERSION}`,
+        `--outdir=${destDir}`,
+        "--log-level=error"
+      )
+      .join(" ");
+    execSync(`npx --yes esbuild next.config.js ${esbuildArgs}  `, { cwd: sourceDir });
```
Comment (jamesdaniels) [inline src/frameworks/next/index.ts:363]: whitespace

File: src/frameworks/next/index.ts [hunk 5]
```diff
@@ -344,25 +344,23 @@ export async function ɵcodegenFunctionsDirectory(sourceDir: string, destDir: st
...
-      bundle: true,
-      external: Object.keys(packageJson.dependencies),
-      absWorkingDir: sourceDir,
-      entryPoints: ["next.config.js"],
-      outfile: join(destDir, "next.config.js"),
-      target: `node${NODE_VERSION}`,
-      platform: "node",
-    });
+    // Bundle their next.config.js with esbuild via NPX, pinned version was having troubles
+    // on m1 macs, older Node versions, and eitherway we should avoid taking on any deps in firebase-tools
```
Comment (jamesdaniels) [inline src/frameworks/next/index.ts:348]: `on m1 mac, older Node version, and eitherway` => `on m1 macs and older Node versions; either way`

File: src/frameworks/next/utils.ts [hunk 6]
```diff
@@ -211,3 +217,16 @@ export function isUsingAppDirectory(dir: string): boolean {
 
   return fileExistsSync(appPathRoutesManifestPath);
 }
+
+/**
+ * Given input from `npm ls` flatten the dependency tree and return all module names
+ *
+ * @param ls JSON returned from `npm ls`
+ */
+export function allDependencyNames(ls: Dependency | null): string[] {
```
Comment (jamesdaniels) [inline src/frameworks/next/utils.ts:226]: Add quick unit test

File: src/frameworks/next/index.ts [hunk 7]
```diff
@@ -344,25 +344,26 @@ export async function ɵcodegenFunctionsDirectory(sourceDir: string, destDir: st
...
-      outfile: join(destDir, "next.config.js"),
-      target: `node${NODE_VERSION}`,
-      platform: "node",
-    });
+    // Bundle their next.config.js with esbuild via NPX, pinned version was having troubles on m1
+    // macs and older Node versions; either way, we should avoid taking on any deps in firebase-tools
+    // Alternatively I tried using @swc/spack and the webpack bundled into Next.js but was
+    // encountering difficulties with both of those
+    const dependencyTree: NpmLsReturn = JSON.parse(
+      execSync(`npm ls --omit=dev --all --json`, { cwd: sourceDir }).toString()
```
Comment (TheIronDev) [inline src/frameworks/next/index.ts:352]: Does this preclude `pNPM` or other package managers? (I honestly don;t know)

**Review by TheIronDev — APPROVED** [review]
Code looks good, but I don't have an M1 mac to test this :(

File: src/frameworks/next/index.ts (hunk 7 above)
Comment (jamesdaniels) [reply src/frameworks/next/index.ts:352]: `npm ls` should still return correct results regardless of the package manager, just confirmed with a codebase of mine that uses yarn

**Comment by leoortizz** [comment]
I successfully deployed a Next.js app with middleware in a M1 Pro using this PR's build

**Pull request merged**